File: StackCalc.py
class Stack:
    def __init__(self):
        self.items = []
        
    def is_empty(self):
        return len(self.items) == 0
    
    def push(self, item):
        self.items.append(item)
        
    def pop(self):
        if self.is_empty():
            return None
        return self.items.pop()
    
    def top(self):
        if self.is_empty():
            return None
        return self.items[-1]
    
    def size(self):
        return len(self.items)

def is_operator(char):
    return char in "+-*/"

def precedence(operator):
    if operator in "+-":
        return 1
    elif operator in "*/":
        return 2
    else:
        return 0

def infix_to_postfix(infix_expression):
    operator_stack = Stack()
    postfix_expression = []
    for char in infix_expression:
        if char.isdigit():
            postfix_expression.append(char)
        elif is_operator(char):
            while (not operator_stack.is_empty()) and precedence(char) <= precedence(operator_stack.top()):
                postfix_expression.append(operator_stack.pop())
            operator_stack.push(char)
        elif char == "(":
            operator_stack.push(char)
        elif char == ")":
            while operator_stack.top() != "(":
                postfix_expression.append(operator_stack.pop())
            operator_stack.pop()
    while not operator_stack.is_empty():
        postfix_expression.append(operator_stack.pop())
    return " ".join(postfix_expression)

def evaluate_postfix(postfix_expression):
    operand_stack = Stack()
    for char in postfix_expression:
        if char.isdigit():
            operand_stack.push(char)
        elif is_operator(char):
            operand2 = operand_stack.pop()
            operand1 = operand_stack.pop()
            result = evaluate(operand1, operand2, char)
            operand_stack.push(result)
    return operand_stack.pop()

def evaluate(operand1, operand2, operator):
    operand1 = int(operand1)
    operand2 = int(operand2)
    if operator == "+":
        return str(operand1 + operand2)
    elif operator == "-":
        return str(operand1 - operand2)
    elif operator == "*":
        return str(operand1 * operand2)
    elif operator == "/":
        return str(operand1 / operand2)

if __name__ == "__main__":
    infix_expression = input("중위 수식을 입력하세요: ")
    postfix_expression = infix_to_postfix(infix_expression)
    result = evaluate_postfix(postfix_expression)

    print("후위 수식: ", postfix_expression)
    print("결과: ", result)

File: StackCalc.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <ctype.h>

#define MAX_SIZE 100

typedef struct stack {
    int top;
    char items[MAX_SIZE];
} Stack;

void push(Stack* stack, char item) {
    if (stack->top == MAX_SIZE - 1) {
        printf("Error: Stack is full\n");
        exit(1);
    }
    stack->items[++stack->top] = item;
}

char pop(Stack* stack) {
    if (stack->top == -1) {
        printf("Error: Stack is empty\n");
        exit(1);
    }
    return stack->items[stack->top--];
}

char top(Stack* stack) {
    if (stack->top == -1) {
        printf("Error: Stack is empty\n");
        exit(1);
    }
    return stack->items[stack->top];
}

int is_operator(char ch) {
    return ch == '+' || ch == '-' || ch == '*' || ch == '/';
}

int precedence(char operator) {
    switch (operator) {
    case '+':
    case '-':
        return 1;
    case '*':
    case '/':
        return 2;
    default:
        return 0;
    }
}

void infix_to_postfix(char* infix_expression, char* postfix_expression) {
    Stack operator_stack;
    operator_stack.top = -1;
    int i = 0, j = 0;

    while (infix_expression[i] != '\0') {
        char ch = infix_expression[i];
        if (isdigit(ch)) {
            postfix_expression[j++] = ch;
        }
        else if (is_operator(ch)) {
            while (operator_stack.top != -1 && precedence(ch) <= precedence(top(&operator_stack))) {
                postfix_expression[j++] = pop(&operator_stack);
            }
            push(&operator_stack, ch);
        }
        else if (ch == '(') {
            push(&operator_stack, ch);
        }
        else if (ch == ')') {
            while (top(&operator_stack) != '(') {
                postfix_expression[j++] = pop(&operator_stack);
            }
            pop(&operator_stack);
        }
        i++;
    }
    while (operator_stack.top != -1) {
        postfix_expression[j++] = pop(&operator_stack);
    }
    postfix_expression[j] = '\0';
}

int evaluate_postfix(char* postfix_expression) {
    Stack operand_stack;
    operand_stack.top = -1;
    int i = 0;

    while (postfix_expression[i] != '\0') {
        char ch = postfix_expression[i];
        if (isdigit(ch)) {
            push(&operand_stack, ch - '0');
        }
        else if (is_operator(ch)) {
            int operand2 = pop(&operand_stack);
            int operand1 = pop(&operand_stack);
            int result = evaluate(operand1, operand2, ch);
            push(&operand_stack, result);
        }
        i++;
    }
    return pop(&operand_stack);
}

int evaluate(int operand1, int operand2, char operator) {
    switch (operator) {
    case '+':
        return operand1 + operand2;
    case '-':
        return operand1 - operand2;
    case '*':
        return operand1 * operand2;
    case '/':
        return operand1 / operand2;
    default:
        return 0;
    }
}

void main() {
    char infix_expression[MAX_SIZE];
    char postfix_expression[MAX_SIZE];
    printf("중위 수식을 입력하세요: ");
    fgets(infix_expression, MAX_SIZE, stdin);
    infix_expression[strcspn(infix_expression, "\n")] = '\0';
    infix_to_postfix(infix_expression, postfix_expression);
    int result = evaluate_postfix(postfix_expression);

    printf("후위 수식: %s\n", postfix_expression);
    printf("결과: %d\n", result);
}
